### ISSUE TITLE
Abstract the the interface of the Client.

### ISSUE THREAD

**Issue by museun** [museun/twitchchat#29]
This will allow for easier integration with AsyncRead/AsyncWrite-style clients, using something like this:
```rust
pub trait Adapter {
    // owned for simplicity but perhaps a &'a [u8] (and use of the `bytes` crate)
    // <= 1024 string ending with \r\n
    fn read_line(&mut self) -> std::io::Result<String>;  
    // <= 512 buffer ending with \r\n
    fn write_line(&mut self, data: &[u8]) -> std::io::Result<usize>; 
}
```

And something like will allow people to use other irc parsing crates to provide the `twitchchat::irc::types::Message` that'll be used with the `twitchchat::Message` types:
```rust
// TODO: perhaps ToOwned / Borrowed (Cow) for these
// @tags :prefix command args :data\r\n
pub trait ToMessage {
  fn tags(&self) -> Option<&str>;
  fn prefix(&self) -> Option<&str>;
  fn command(&self) -> Option<&str>;
  fn args(&self) -> Option<&[&str]>;
  fn data(&self) -> Option<&str>;
}

// twitchchat::Message
impl Message {
  fn parse(msg: impl ToMessage) -> Self {
    let tags = msg.tags();
    let prefix = msg.prefix();
    let command = msg.command();
    let args = msg.args();
    let data = msg.data();

    // do parsing here
  }
}
```
(which will allow for removing the whole `irc` module and providing an optional one that implements this trait)

**Comment by museun** [comment]
Also, maybe split the Adapter to allow for variations of Send/Sync (!Send/!Sync). The `writer` portion of it should be clonable and/or thread safe, but it should be up to the user. The `reader` should probably not be clonable, nor thread safe though. (simplification of the filter/callback thing). 

This is under the assumption most consumers of this crate want to react (e.g. replying as a bot) to messages more than they want to just read them.

**Comment by museun** [comment]
closed with b2f52b0
